### ISSUE TITLE
[dotnet] Bring back Dispose optimization

### ISSUE THREAD

**Issue by spouliot** [xamarin/xamarin-macios#11590]
# Problem

Backing fields for weak properties are not removed by ILLink.

## Why ? 

When the instance `Handle` is null (`IntPtr.Zero`) then the generated `Dispose` method sets all backing fields to `null`.

This is a problem for

Linker
* `Dispose` is overridden and required for `IDisposable` so it will **always** be marked for `NSObject` subclasses
* `Dispose` reference each backing field - which, in turn, which marks all of the backing fields

Performance
* Each backing field adds a pointer-size (8 bytes) to each instance of the type. Memory usage for heavy UI apps is impacted.

# Existing solution (Legacy)

## Postpone and Rewrite

* The original linker `MarkStep` is subclassed
* Processing `Dispose` for `NSObject` subclasses is delayed to see which fields are marked (by something else than `Dispose`).
* When the processing is completed then the `Dispose` are re-written to only use the marked backing fields. The ones that were not marked (not used) are then _naturally_ removed by the linker.

## Pros
* It's been working for years (with the old linker code base)
* Eliminate unused backing fields (important for memory footprint)

## Cons
* We cannot subclass `MarkStep` anymore and are missing API to do something similar
* Not 100% optimal, we can end up with a chain of nearly empty `Dispose`

```csharp
protected override void Dispose(bool disposing)
{
	base.Dispose(disposing);
	if (base.Handle == IntPtr.Zero)
	{
	}
}
```

In part because C# general-looking `base.Dispose` call is in fact
```cil
call instance void Foundation.NSObject::Dispose(bool)
```
IOW it call a specific subclass and removing an intermediate `Dispose` breaks the chain. However it was never identified as a performance issue (nor had a huge size impact) so this was never fully optimized.

## Notes

* To avoid marking additional types all backing fields are generated as `System.Object`

# Proposal for .net

With ILLink we do not have access to the API to re-implement the same solution. A different approach is needed.

## Use reflection to `null` the backing fields

FWIW I'm both excited and horrified by the idea :|

## Pros
* No changes are required for the linker code / steps
	* Good for build performance
* Only a single method `Dispose` method is needed, it can (and should be) inside `NSObject`
	* Less code (at least a fixed amount), less overrides, less metadata
* Linker won't mark the fields, unless they are set inside properties (and are needed)
	* because `Dispose` won't reference any of them (as there's no generated `Dispose` in subclasses)
* Object's instance size are optimal (like legacy)

## Cons
* Requires a small generator changes (skip `Dispose` overrides)
* Runtime performance would suffer (needs measurements) but...

If needed we can re-generate the `Dispose` methods, as a (release) optimization, in IL. As such it rarely has to run (and not when build performance is critical).

That would lose some of the pros (and should be optional depending if you prefer performance over size for your release build). That would still be better than legacy as there would be an optimal/minimal (so less _empty_) number of `Dispose` method re-created.

Unless the impact is easily noticeable the optimization can also be postponed to a later time. We still get the size reduction and memory footprint gains easily.

## Gotchas

* Field names must **not** be changed, e.g. metadata reducer (not yet enabled), before the processing is over (or when reflection is used).

## Alternative (.net)

An alternative approach would be to:

* Identify all `NSObject` subclasses [p1] before running the linker (Mark step)
	* Keep a list of all backing fields for each type
	* Re-write `Dispose` so it does not have a reference to any of them (`nop` stuff)
* Mark and Sweep (as normal)
* For each `NSObject` subclasses that was marked
	* Restore the code (set to null) only for the marked backing fields
* Save assemblies (as normal)

Code to remove/add back for each field looks like
```cil
ldarg.0
ldnull
stfld object UIKit.UIViewController::__mt_XXX_var
```

## Pros
* Performance is **always** identical (should be identical to legacy, at least the IL would be)
	* Reflection is never used
	* But not optimal, there are more overridden calls (same as legacy)

## Cons
* Not as efficient for size
	* Only one `Dispose` needed for size-optimal case (reflection)
	* We would get extraneous `Dispose` methods (like legacy) - ones that are empty beside calling `base`
* Not best performance (but close)
	* Best performance would be to re-write all `Dispose` from scratch which could still be done for this scenario, just like it could have been done for legacy. Same question: is it worth the time investment ?

## Problem

* [p1] we do not have full control over the loading of assemblies, e.g. linking can start with assemblies not yet loaded. This is not the only place where this is a problem so a solution might be shared to address this.

**Comment by spouliot** [comment]
# Use reflection to null the backing fields

The "release optimizer" for this scenario does not seems too complex (less than the original linker optimization). We get can notified on marks, so we can track the backing fields/types that are required.

From there we can generate the (missing) dispose methods, chaining the `base.Dispose` requires a bit more work (a tree for NSObject _marked_ subclasses).

An we can remove the reflection code from `NSObject.Dispose` - unless there are NSObject subclasses that are not linked (so we can't add code) and have backing fields.

In the case of our "minimal" sample we would not generate any new `Dispose` (no backing fields are needed) but the reflection "cost" (that hides the benefits for such a small test case) would be removed.

**Comment by marek-safar** [comment]
I think it would make sense to fix this in the linker and don't mark fields that are only assigned with primitive or null values.

**Comment by spouliot** [comment]
@marek-safar I'm not opposed to have the linker do this but it's not as simple as just avoiding marking some field.

```csharp

private object __mt_ParentFocusEnvironment_var;
private object __mt_PreferredFocusedView_var;
private object __mt_WeakTransitioningDelegate_var;

protected override void Dispose(bool disposing)
{
	base.Dispose(disposing);
	if (base.Handle == IntPtr.Zero)
	{
 		__mt_ParentFocusEnvironment_var = null;
		__mt_PreferredFocusedView_var = null;
		__mt_WeakTransitioningDelegate_var = null;
	}
}
```

In the above code it's possible that `__mt_ParentFocusEnvironment_var` is needed, while the 2 other are not.

If you only mark `__mt_ParentFocusEnvironment_var` then we get ride of `__mt_PreferredFocusedView_var` and  `__mt_WeakTransitioningDelegate_var ` fields (good) but the `Dispose` method becomes invalid (bad).

To fix this you have to re-write the IL to remove references to the unmarked fields. 

It's not hard (we have this in legacy) but it needs to be done once marking is complete since the linker could process `Dispose` before any of code using the fields are processed.

**Comment by marek-safar** [comment]
> To fix this you have to re-write the IL to remove references to the unmarked fields.

Sure, the linker would do that part too. However, it would not remove `if (base.Handle == IntPtr.Zero)` condition (it could but that's more tricky) so the method override would stay.

**Comment by spouliot** [comment]
> However, it would not remove if (base.Handle == IntPtr.Zero) condition (it could but that's more tricky) so the method override would stay.

That's fine with me as this is what we have today with the (legacy) SDK. 

Not optimal but never important enough to fix. We can re-assess that in the (far) future :)

**Comment by Therzok** [comment]
These fields cause more problems and headaches than they're worth, especially since Dispose is usually not a safe operation.

**Comment by spouliot** [comment]
A different approach was taken with fccd43f0c224c6bb48cb63524b408f21047795cb

if/when ILLink supports this then our custom mark handler / step can be removed
